### PR TITLE
feat: implement getMessageReceipt and tests

### PR DESCRIPTION
**Pull request by smartcontracts**
**Description**
Implements the getMessageReceipt function and corresponding tests. Also adds a new `toCrossChainMessage` function that can parse a `MessageLike` into a `CrossChainMessage` object.

**Metadata**
- Fixes ENG-1800

### CODE REVIEW

**Comment by tynes** [comment]
With all of the todos, is this a draft or ready for review?

**Comment by smartcontracts** [comment]
> With all of the todos, is this a draft or ready for review?

The TODOs are mainly just notes for myself in the future, nothing blocking this PR. Going to run over all of the TODOs later and make tickets out of them if they're actually worth doing.

File: packages/sdk/src/cross-chain-provider.ts [hunk 1]
```diff
@@ -278,14 +280,110 @@ export class CrossChainProvider implements ICrossChainProvider {
...
+        message as TransactionLike
+      )
+
+      // TODO: Can we do better verification here?
+      const found = messages
+        .sort((a, b) => {
+          return a.logIndex - b.logIndex
+        })
+        .find((m) => {
+          return m.logIndex > (message as TokenBridgeMessage).logIndex
```
Comment (tynes) [inline packages/sdk/src/cross-chain-provider.ts:304]: Should this be `===` ?

File: packages/sdk/src/utils/misc-utils.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// TODO: A lot of this stuff could probably live in core-utils instead.
+// Review this file eventually for stuff that could go into core-utils.
+export const omit = (obj: any, ...keys: string[]) => {
+  return keys.reduce((acc, key) => {
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const { [key]: omitted, ...rest } = acc
```
Comment (tynes) [inline packages/sdk/src/utils/misc-utils.ts:6]: I'm confused here - is `omitted` defined? Also I think this could be implemented more cleanly by copying `obj` and then iterating over keys and deleting each one using `delete`

File: packages/sdk/src/cross-chain-provider.ts (hunk 1 above)
Comment (smartcontracts) [reply packages/sdk/src/cross-chain-provider.ts:304]: We need to find the first SentMessage event such that the logIndex is greater than the TokenBridgeMessage logIndex. Basically we detect TokenBridgeMessage events separately (they're usually stuff like `ERC20DepositInitiated` or `WithdrawalInitiated`). These events are then followed by a SentMessage event.

File: packages/sdk/src/utils/misc-utils.ts (hunk 2 above)
Comment (smartcontracts) [reply packages/sdk/src/utils/misc-utils.ts:6]: Yeah I guess this is personal preference. Either approach works here (`delete` or destructuring). `omitted` is undefined and just used to pull the variable out.

Comment (smartcontracts) [reply packages/sdk/src/utils/misc-utils.ts:6]: I'll edit this to use `delete`, I think that's more straightforward

Comment (smartcontracts) [reply packages/sdk/src/utils/misc-utils.ts]: Replaced to use `delete`. Resolving this comment.

File: packages/sdk/src/cross-chain-provider.ts (hunk 1 above)
Comment (smartcontracts) [reply packages/sdk/src/cross-chain-provider.ts:304]: I can add a comment here to make this more clear

Comment (smartcontracts) [reply packages/sdk/src/cross-chain-provider.ts:311]: I added a detailed comment that explains what's happening here

File: packages/sdk/src/cross-chain-provider.ts [hunk 3]
```diff
@@ -278,14 +280,119 @@ export class CrossChainProvider implements ICrossChainProvider {
...
+    // TODO: Convert these checks into proper type checks.
+    if ((message as CrossChainMessage).message) {
+      return message as CrossChainMessage
+    } else if (
+      (message as TokenBridgeMessage).l1Token &&
+      (message as TokenBridgeMessage).l2Token &&
+      (message as TokenBridgeMessage).transactionHash
+    ) {
+      const messages = await this.getMessagesByTransaction(
+        message as TransactionLike
```
Comment (smartcontracts) [inline packages/sdk/src/cross-chain-provider.ts:295]: Use transactionHash directly

File: packages/sdk/src/cross-chain-provider.ts [hunk 4]
```diff
@@ -278,14 +280,119 @@ export class CrossChainProvider implements ICrossChainProvider {
...
+    // enough to trigger the event. However, for now we need a temporary way to find L1 => L2
+    // transactions that fail but don't alert us because they didn't provide enough gas.
+    // TODO: Talk with the systems and protocol team about coordinating a hard fork that fixes this
+    // on both L1 and L2.
+
+    // Just return null if we didn't find a receipt. Slightly nicer than throwing an error.
+    return null
   }
 
   public async waitForMessageReciept(
```
Comment (SmitV) [inline packages/sdk/src/cross-chain-provider.ts:398]: Not part of the PR but I before E except after C :)

**Review by SmitV — APPROVED** [review]
Approving based on review on call. Good stuff
